libvast/src/system/type_registry.cpp

```diff
@@ -88,11 +88,48 @@ caf::error type_registry_state::load_from_disk() {
 }
 
 void type_registry_state::insert(vast::type layout) {
-  auto& cont = data[layout.name()].value;
+  auto flattened_layout = flatten(layout);
+  auto& old_layouts = data[flattened_layout.name()].value;
+  // Helper function.
+  auto is_compatible = [](const type& x) noexcept {
+    auto lhs = caf::get_if<record_type>(&x);
+    return [&, lhs](const type& y) noexcept {
+      auto rhs = caf::get_if<record_type>(&y);
+      // If they are not record types, check if they are congruent.
+      if (!lhs || !rhs)
+        return congruent(x, y);
+      // Check whether all fields of rhs exist in lhs, i.e., no new fields were
+      // added.
+      for (auto&& rf : rhs->fields) {
+        if (auto lf = lhs->find(rf.name)) {
+          // If there's a type mismatch, exit early as well.
+          if (!congruent(lf->type, rf.type))
+            return false;
+        } else {
+          // Not all fields of rhs exist in lhs. Exit early.
+          return false;
+        }
+      }
+      // We passed the test for all fields. Yay!
+      return true;
+    };
+  };
+  // Check whether the new layout is compatible with the old one. If it isn't,
+  // forget about old versions of the layout.
+  auto are_all_compatible = std::all_of(old_layouts.begin(), old_layouts.end(),
+                                        is_compatible(flattened_layout));
+  if (!are_all_compatible) {
+    VAST_VERBOSE(self, "detected an incompatible version of",
```

Comment:
```suggestion
    VAST_WARNING(self, "detected an incompatible version of",
```

Up to you, but I'd see this as something beyond VERBOSE.

Reply:
Fine by me.